## Get last 3 months of articles sales and store in file

We want to have a dataframe with the average monthly units sold from July, August and September 2018 and the units sold in October 2018 to use as a benchmark to help test our model results. 

## Notebook setup

Import packages and load data

In [5]:
# Navigate to directory with cleaned data
#%cd /data/p_dsi/teams2023/team7

/gpfs52/data/p_dsi/teams2023/team7


In [70]:
import pandas as pd
import numpy as np
import seaborn as sns
#from janitor import clean_names
from sklearn.preprocessing import OneHotEncoder

In [3]:
pd.options.display.max_columns = 100

In [57]:
# Read-in data
sales = pd.read_csv("final_sales.csv", parse_dates=["DATE"], 
                    usecols=["ARTICLE_ID","DATE","UNITS"], 
                    dtype={"ARTICLE_ID":"category","UNITS":np.float64},
                    low_memory=False)

# (1) Store shape, (2) total sum of units and (3) 2018 total sum of units for sanity checks
# 1
sales_shape = sales.shape
# 2
total_units_sold = sales["UNITS"].sum()
#3
sales = sales.set_index("DATE")
total_units_sold_in_2018 = sales.loc['2018']['UNITS'].sum()

# Print completion message
print("Data read-in complete. \nShape of dataframe: ", sales.shape, "\n Total unit sales across all articles: ", total_units_sold)

Data read-in complete. 
Shape of dataframe:  (13499278, 2) 
 Total unit sales across all articles:  30299003.0


In [59]:
sales.head(3)

ARTICLE_ID  UNITS
DATE                        
2018-07-22       3431    1.0
2018-07-05       3448    1.0
2018-07-02     147127    1.0

## Yearly tire sales:

In [63]:
# Information about the articles sold
print("\n Total units sold in 2018 across all articles: ", get_total_units_sold(sales, 2018))
print("\n Total units sold in 2017 across all articles: ", get_total_units_sold(sales, 2017))
print("\n Total units sold in 2016 across all articles: ", get_total_units_sold(sales, 2016))
print("\n Total units sold in 2015 across all articles: ", get_total_units_sold(sales, 2015))


 Total units sold in 2018 across all articles:  7907212.0

 Total units sold in 2017 across all articles:  9008568.0

 Total units sold in 2016 across all articles:  8191490.0

 Total units sold in 2015 across all articles:  5191733.0


In [12]:
# Sort
sales.sort_values(by=["DATE","ARTICLE_ID"], inplace=True)

# Group the rows by ARTICLE_ID and month of DATE, and sum the UNITS values
grouped_sales = 

# Create new month and year columns
grouped_sales['MONTH'] = grouped_sales['DATE'].dt.month
grouped_sales['YEAR'] = grouped_sales['DATE'].dt.year
grouped_sales = grouped_sales.drop(["DATE"],axis=1)

print("Shape of grouped_sales: ", grouped_sales.shape)
print("Simple average of monthly units sold per article: ", grouped_sales.shape[0]/420, "\n")

print("Some stats: ", grouped_sales.describe(), "\n")

print(grouped_sales.sample(10, random_state=1))

Shape of grouped_sales:  (18705, 4)
Simple average of monthly units sold per article:  44.535714285714285 

Some stats:                UNITS         MONTH          YEAR
count  18705.000000  18705.000000  18705.000000
mean    1619.834429      6.581395   2016.534884
std     1910.748041      3.265356      1.064220
min       -2.000000      1.000000   2015.000000
25%      452.000000      4.000000   2016.000000
50%     1000.000000      7.000000   2017.000000
75%     2130.000000      9.000000   2017.000000
max    23338.000000     12.000000   2018.000000 

      ARTICLE_ID   UNITS  MONTH  YEAR
11602       3928   701.0      3  2018
4785       15029  6789.0      4  2016
3581        1428     6.0      4  2016
11999       4022     0.0      6  2015
14920       6410   867.0     10  2018
15295       6896   548.0     10  2017
8126        2360  1020.0     10  2018
7160        2012   781.0      2  2017
3552        1427  3040.0      6  2017
14417       6240   623.0      4  2016


In [13]:
# Sanity check 1 PASS
print("Sum of units in grouped_sales: ", sum(grouped_sales['UNITS']))
print("Sum of units sold before grouping transformation: ", total_units_sold)

# Sanity check 2
grouped_sales_2018 = grouped_sales[grouped_sales['YEAR'] == 2018]['UNITS'].sum()
print("Sum of units sold in 2018 in grouped_sales: ", grouped_sales_2018)
print("Sum of units sold in 2018 before grouping transformation: ", total_units_sold_in_2018)


Sum of units in grouped_sales:  30299003.0
Sum of units sold before grouping transformation:  30299003.0
Sum of units sold in 2018 in grouped_sales:  7907212.0
Sum of units sold in 2018 before grouping transformation:  7907212.0


In [14]:
# pivot the dataframe
sales_pivoted = grouped_sales.pivot(index=['YEAR', 'ARTICLE_ID'], columns='MONTH', values='UNITS').reset_index()

# print the pivoted dataframe
sales_pivoted

MONTH  YEAR ARTICLE_ID       1       2       3       4       5       6   
0      2015     106242     NaN     NaN     NaN   149.0   147.0   126.0  \
1      2015     106259     NaN     NaN     NaN   869.0  1096.0  1185.0   
2      2015     106310     NaN     NaN     NaN   969.0  1435.0  1470.0   
3      2015     106497     NaN     NaN     NaN  4011.0  3735.0  4022.0   
4      2015     106650     NaN     NaN     NaN   734.0   723.0   976.0   
...     ...        ...     ...     ...     ...     ...     ...     ...   
1735   2018      98031   559.0   457.0   418.0   381.0   402.0   553.0   
1736   2018      98048   767.0   722.0  1037.0   782.0   951.0  1002.0   
1737   2018      98065  2237.0  2250.0  2384.0  2153.0  2197.0  2604.0   
1738   2018      98099  2490.0  2343.0  2700.0  2352.0  2703.0  3224.0   
1739   2018      98150  2526.0  2393.0  2251.0  2216.0  2565.0  3225.0   

MONTH       7       8       9      10      11      12  
0       144.0   132.0   136.0   212.0   209.0   210.0  
1      1278.0  1122.0  1239.0  1029.0  1486.0  1253.0  
2      1694.0  1938.0  2140.0  1752.0  1885.0  1436.0  
3      3808.0  4941.0  3136.0  2640.0  3981.0  3034.0  
4      1053.0  1507.0  1770.0  1293.0  1549.0  1017.0  
...       ...     ...     ...     ...     ...     ...  
1735    408.0   410.0   321.0   295.0     NaN     NaN  
1736    979.0   922.0   733.0   653.0     NaN     NaN  
1737   2457.0  2799.0  2478.0  1887.0     NaN     NaN  
1738   2986.0  2823.0  2929.0  3042.0     NaN     NaN  
1739   3131.0  2942.0  2842.0  3082.0     NaN     NaN  

[1740 rows x 14 columns]

In [15]:

temp = sales_pivoted[['ARTICLE_ID','YEAR', 7, 8, 9, 10]]
temp['AvgLast3Months'] = temp[[7, 8, 9]].apply(lambda x: x.mean(), axis=1)
temp = temp[['ARTICLE_ID','YEAR', 7, 8, 9,'AvgLast3Months',10]]
final = temp[temp['YEAR'] == 2018]
final

/var/folders/jc/57b2n61s3sgd__mz5yq9vvmc0000gn/T/ipykernel_52189/1988501536.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp['AvgLast3Months'] = temp[[7, 8, 9]].apply(lambda x: x.mean(), axis=1)


MONTH ARTICLE_ID  YEAR       7       8       9  AvgLast3Months      10
1305      106242  2018   431.0   552.0   564.0      515.666667   507.0
1306      106259  2018   709.0   722.0   645.0      692.000000   687.0
1307      106310  2018  1059.0  1024.0   534.0      872.333333   417.0
1308      106497  2018  1457.0  1713.0  1765.0     1645.000000  1412.0
1309      106650  2018   555.0   676.0   743.0      658.000000   573.0
...          ...   ...     ...     ...     ...             ...     ...
1735       98031  2018   408.0   410.0   321.0      379.666667   295.0
1736       98048  2018   979.0   922.0   733.0      878.000000   653.0
1737       98065  2018  2457.0  2799.0  2478.0     2578.000000  1887.0
1738       98099  2018  2986.0  2823.0  2929.0     2912.666667  3042.0
1739       98150  2018  3131.0  2942.0  2842.0     2971.666667  3082.0

[435 rows x 7 columns]

In [16]:
quant = sales_pivoted.query('YEAR == 2017')[11].quantile(0.9)

sales_pivoted[sales_pivoted[11] >= quant].query('YEAR == 2017').sort_values(by=11, ascending=False)

## TODO - check end of this notebook and around here 
## Want to have a notebook going that addresses articles with the biggest gaps between Oct-Nov and Sept-Nov
## because those articles will really hurt our score. 
## Also note issue with articles that are newer and so have harder sales patterns to predict

MONTH  YEAR ARTICLE_ID        1        2        3        4        5        6   
1296   2017      97946   7344.0   6237.0   7825.0   7617.0   7659.0   8163.0  \
1294   2017      97912   7209.0   6858.0   9239.0   8319.0   8317.0   9129.0   
886    2017     114674   7738.0  10347.0  14452.0  12580.0  12806.0  11262.0   
1287   2017      97759   8163.0   7318.0   8219.0   7749.0   7689.0   8017.0   
911    2017     123888   6152.0   6651.0  12303.0  10001.0  10806.0  10125.0   
1208   2017       6251      0.0      0.0      0.0      0.0      0.0      0.0   
903    2017     122596   6873.0   7638.0  10878.0   9652.0   9914.0   8831.0   
1234   2017    7099618   7665.0   7209.0   8663.0   8358.0   8865.0   9196.0   
986    2017      15131   8230.0   6973.0   7584.0   6825.0   8243.0   8462.0   
987    2017      15148   8748.0   8074.0   7800.0   7413.0   8795.0   9273.0   
1003   2017      15505   7282.0   7242.0   7471.0   6579.0   7889.0   8112.0   
1293   2017      97895   6228.0   5269.0   7277.0   6447.0   6647.0   7131.0   
1299   2017      97997   5212.0   4838.0   6007.0   5517.0   5783.0   6161.0   
1134   2017       3816      0.0      0.0     12.0     12.0     47.0     44.0   
1105   2017       3431  15528.0  13509.0   7902.0  10269.0   8757.0   9367.0   
1254   2017        832   6454.0   7716.0   9625.0   7339.0   7317.0   6529.0   
905    2017     122630   4243.0   5421.0   8497.0   7285.0   6930.0   6760.0   
1209   2017       6252      0.0      0.0      0.0      0.0      0.0      0.0   
920    2017     136094   4804.0   4377.0   5120.0   5224.0   5736.0   6115.0   
1103   2017       3429  13442.0   7186.0   2050.0   6334.0   6090.0   7848.0   
904    2017     122613   5470.0   5592.0   7384.0   7098.0   7006.0   6403.0   
883    2017     114572   6415.0   6778.0   8630.0   6858.0   7253.0   7208.0   
1285   2017      97725   5531.0   4663.0   5662.0   5830.0   5300.0   5510.0   
1239   2017    7099718   5244.0   5253.0   6076.0   5532.0   6167.0   6223.0   
1248   2017        826   5549.0   4729.0   5461.0   6037.0   6224.0   6215.0   
1297   2017      97963   3386.0   2819.0   3674.0   3900.0   3869.0   4362.0   
1238   2017    7099717   4542.0   4761.0   5731.0   5116.0   5457.0   5439.0   
1109   2017       3438   5416.0   6879.0   4351.0   7191.0   6490.0   6161.0   
1130   2017       3724   5415.0   6098.0   7190.0   5932.0   6754.0   6563.0   
977    2017      14944   4535.0   4217.0   3834.0   3435.0   5746.0   6394.0   
1290   2017      97827   3353.0   3307.0   4187.0   4090.0   3828.0   4451.0   
938    2017       1408   2837.0   2741.0   3055.0   2912.0   3445.0   3504.0   
1129   2017       3707   3867.0   5043.0   5877.0   5254.0   5757.0   5969.0   
912    2017     123905   4629.0   4776.0   5129.0   4780.0   5164.0   5045.0   
952    2017       1427   2510.0   1901.0   3048.0   3338.0   2443.0   3040.0   
884    2017     114589   3605.0   3591.0   5450.0   4205.0   4433.0   4165.0   
1221   2017      66105   3117.0   3197.0   3968.0   3141.0   3916.0   4165.0   
1030   2017     189769   3647.0   3083.0   3918.0   3644.0   4343.0   4531.0   
940    2017       1411   2223.0   2061.0   2785.0   2360.0   3516.0   3436.0   
1264   2017      92234   3881.0   4008.0   5039.0   4380.0   4748.0   4519.0   
1186   2017      53967   3118.0   3888.0   4709.0   4344.0   4267.0   4745.0   
1149   2017       4022      0.0      0.0      1.0     50.0     77.0     82.0   
1298   2017      97980   3106.0   3041.0   3685.0   3592.0   3624.0   3980.0   
1232   2017    7099616   3445.0   3311.0   4191.0   4026.0   4282.0   4202.0   

MONTH        7        8        9       10       11       12  
1296    8821.0   8367.0   9757.0   8774.0  12071.0  10569.0  
1294    9493.0   8201.0   9110.0   9872.0  10899.0   9238.0  
886    14099.0  14687.0  11148.0  12606.0  10753.0   9362.0  
1287    9362.0   8580.0   8995.0   8744.0  10475.0   9114.0  
911    12770.0  13273.0  11226.0  11556.0   9956.0   8095.0  
1208       0.0      0.0    

In [17]:
temp = sales_pivoted[['ARTICLE_ID','YEAR', 7, 8, 9, 10]]
temp['AvgLast3Months'] = temp[[7, 8, 9]].apply(lambda x: x.mean(), axis=1)
temp = temp[['ARTICLE_ID','YEAR', 7, 8, 9,'AvgLast3Months',10]]
final = temp[temp['YEAR'] == 2018]
final

/var/folders/jc/57b2n61s3sgd__mz5yq9vvmc0000gn/T/ipykernel_52189/1988501536.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp['AvgLast3Months'] = temp[[7, 8, 9]].apply(lambda x: x.mean(), axis=1)


MONTH ARTICLE_ID  YEAR       7       8       9  AvgLast3Months      10
1305      106242  2018   431.0   552.0   564.0      515.666667   507.0
1306      106259  2018   709.0   722.0   645.0      692.000000   687.0
1307      106310  2018  1059.0  1024.0   534.0      872.333333   417.0
1308      106497  2018  1457.0  1713.0  1765.0     1645.000000  1412.0
1309      106650  2018   555.0   676.0   743.0      658.000000   573.0
...          ...   ...     ...     ...     ...             ...     ...
1735       98031  2018   408.0   410.0   321.0      379.666667   295.0
1736       98048  2018   979.0   922.0   733.0      878.000000   653.0
1737       98065  2018  2457.0  2799.0  2478.0     2578.000000  1887.0
1738       98099  2018  2986.0  2823.0  2929.0     2912.666667  3042.0
1739       98150  2018  3131.0  2942.0  2842.0     2971.666667  3082.0

[435 rows x 7 columns]

In [18]:
final.to_csv("last_3_mo_articles_sold.csv", index=False)

In [19]:
def get_wmape(df, actuals='actual', predictions='prediction'):
  '''
  Function takes a dataframe with two columns for actual and predicted values.
  actuals = string of the column name with actual values
  predictions = string of the column name with predicted values
  Returns the WMAPE score.
  '''
  return sum(abs(df[actuals]-df[predictions]))/sum(df[actuals])

In [20]:
get_wmape(final, actuals=10, predictions='AvgLast3Months')

0.13183069311862147

## WMAPE for average most recent 2 months

In [21]:
final['AvgLast2Months'] = final[[8, 9]].apply(lambda x: x.mean(), axis=1)
final

/var/folders/jc/57b2n61s3sgd__mz5yq9vvmc0000gn/T/ipykernel_52189/2864211015.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final['AvgLast2Months'] = final[[8, 9]].apply(lambda x: x.mean(), axis=1)


MONTH ARTICLE_ID  YEAR       7       8       9  AvgLast3Months      10   
1305      106242  2018   431.0   552.0   564.0      515.666667   507.0  \
1306      106259  2018   709.0   722.0   645.0      692.000000   687.0   
1307      106310  2018  1059.0  1024.0   534.0      872.333333   417.0   
1308      106497  2018  1457.0  1713.0  1765.0     1645.000000  1412.0   
1309      106650  2018   555.0   676.0   743.0      658.000000   573.0   
...          ...   ...     ...     ...     ...             ...     ...   
1735       98031  2018   408.0   410.0   321.0      379.666667   295.0   
1736       98048  2018   979.0   922.0   733.0      878.000000   653.0   
1737       98065  2018  2457.0  2799.0  2478.0     2578.000000  1887.0   
1738       98099  2018  2986.0  2823.0  2929.0     2912.666667  3042.0   
1739       98150  2018  3131.0  2942.0  2842.0     2971.666667  3082.0   

MONTH  AvgLast2Months  
1305            558.0  
1306            683.5  
1307            779.0  
1308           1739.0  
1309            709.5  
...               ...  
1735            365.5  
1736            827.5  
1737           2638.5  
1738           2876.0  
1739           2892.0  

[435 rows x 8 columns]

In [22]:
get_wmape(final, actuals=10, predictions='AvgLast2Months')

0.1187478107079956

In [23]:
final['AbsDiff_Last2'] = abs(final[10] - final['AvgLast2Months'])
final['AbsDiff_Last3'] = abs(final[10] - final['AvgLast3Months'])

final['AbsDiff_Last2'].sum(), final['AbsDiff_Last3'].sum()

/var/folders/jc/57b2n61s3sgd__mz5yq9vvmc0000gn/T/ipykernel_52189/218765272.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final['AbsDiff_Last2'] = abs(final[10] - final['AvgLast2Months'])
/var/folders/jc/57b2n61s3sgd__mz5yq9vvmc0000gn/T/ipykernel_52189/218765272.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final['AbsDiff_Last3'] = abs(final[10] - final['AvgLast3Months'])


(96954.5, 107636.33333333334)

In [100]:
final

MONTH  level_0  index ARTICLE_ID  YEAR       7       8       9   
0            0   1305     106242  2018   431.0   552.0   564.0  \
1            1   1306     106259  2018   709.0   722.0   645.0   
2            2   1307     106310  2018  1059.0  1024.0   534.0   
3            3   1308     106497  2018  1457.0  1713.0  1765.0   
4            4   1309     106650  2018   555.0   676.0   743.0   
..         ...    ...        ...   ...     ...     ...     ...   
430        430   1735      98031  2018   408.0   410.0   321.0   
431        431   1736      98048  2018   979.0   922.0   733.0   
432        432   1737      98065  2018  2457.0  2799.0  2478.0   
433        433   1738      98099  2018  2986.0  2823.0  2929.0   
434        434   1739      98150  2018  3131.0  2942.0  2842.0   

MONTH  AvgLast3Months      10  AvgLast2Months  AbsDiff_Last2  AbsDiff_Last3  
0          515.666667   507.0           558.0           51.0       8.666667  
1          692.000000   687.0           683.5            3.5       5.000000  
2          872.333333   417.0           779.0          362.0     455.333333  
3         1645.000000  1412.0          1739.0          327.0     233.000000  
4          658.000000   573.0           709.5          136.5      85.000000  
..                ...     ...             ...            ...            ...  
430        379.666667   295.0           365.5           70.5      84.666667  
431        878.000000   653.0           827.5          174.5     225.000000  
432       2578.000000  1887.0          2638.5          751.5     691.000000  
433       2912.666667  3042.0          2876.0          166.0     129.333333  
434       2971.666667  3082.0          2892.0          190.0     110.333333  

[435 rows x 12 columns]

## Research: Which months are good predictors for November, and which are not? 

In [25]:
# Group the rows by ARTICLE_ID and month of DATE, and sum the UNITS values
grouped_sales = sales.groupby(['ARTICLE_ID', pd.Grouper(key='DATE', freq='M')])['UNITS'].sum().reset_index()

# Create new month and year columns
grouped_sales['MONTH'] = grouped_sales['DATE'].dt.month
grouped_sales['YEAR'] = grouped_sales['DATE'].dt.year
grouped_sales = grouped_sales.drop(["DATE"],axis=1)

# Filter for 2017
grouped_sales2017 = grouped_sales[grouped_sales['YEAR'] == 2017]
sales_pivoted2017 = grouped_sales2017.pivot(index=['YEAR', 'ARTICLE_ID'], columns='MONTH', values='UNITS').reset_index()

# Add comparison columns
sales_pivoted2017['Oct-Nov'] = sales_pivoted2017[11]-sales_pivoted2017[10]
sales_pivoted2017['Sept-Oct'] = sales_pivoted2017[10]-sales_pivoted2017[9]
sales_pivoted2017['Perc_Oct-Nov'] = round((sales_pivoted2017[11]-sales_pivoted2017[10])/sales_pivoted2017[10]*100,1)
sales_pivoted2017['Perc_Sept-Oct'] = round((sales_pivoted2017[10]-sales_pivoted2017[9])/sales_pivoted2017[9]*100,1)
sales_pivoted2017['AvgSeptOct'] = sales_pivoted2017[[9, 10]].apply(lambda x: x.mean(), axis=1)
sales_pivoted2017['Nov-AvgSeptOct'] = sales_pivoted2017[11]-sales_pivoted2017['AvgSeptOct']
sales_pivoted2017['Abs(Nov-AvgSeptOct)'] = abs(sales_pivoted2017['Nov-AvgSeptOct'])
sales_pivoted2017[5:11]


MONTH  YEAR ARTICLE_ID       1      2      3       4       5       6       7   
5      2017     108860   441.0  431.0  530.0   503.0   455.0   478.0   468.0  \
6      2017     108894  1016.0  810.0  997.0  1025.0  1229.0  1067.0  1007.0   
7      2017     108911   575.0  550.0  608.0   596.0   632.0   675.0   698.0   
8      2017      11187   640.0  657.0  736.0   757.0   739.0   759.0   792.0   
9      2017     112345   752.0  705.0  951.0   717.0  1472.0  1303.0  1682.0   
10     2017        113     0.0    0.0    0.0     0.0     0.0     0.0     0.0   

MONTH       8       9      10      11      12  Oct-Nov  Sept-Oct   
5       574.0   664.0   598.0   657.0   763.0     59.0     -66.0  \
6      1080.0  1330.0  1623.0  1940.0  1637.0    317.0     293.0   
7       803.0   905.0  1021.0  1165.0  1049.0    144.0     116.0   
8       679.0   582.0   671.0   492.0   601.0   -179.0      89.0   
9      2113.0  3087.0  3036.0  2998.0  2553.0    -38.0     -51.0   
10        0.0     0.0    27.0    52.0    78.0     25.0      27.0   

MONTH  Perc_Oct-Nov  Perc_Sept-Oct  AvgSeptOct  Nov-AvgSeptOct   
5               9.9           -9.9       631.0            26.0  \
6              19.5           22.0      1476.5           463.5   
7              14.1           12.8       963.0           202.0   
8             -26.7           15.3       626.5          -134.5   
9              -1.3           -1.7      3061.5           -63.5   
10             92.6            inf        13.5            38.5   

MONTH  Abs(Nov-AvgSeptOct)  
5                     26.0  
6                    463.5  
7                    202.0  
8                    134.5  
9                     63.5  
10                    38.5

In [26]:
sales_pivoted2017.sort_values(['Abs(Nov-AvgSeptOct)'], ascending=False).head(20)

MONTH  YEAR ARTICLE_ID       1        2        3        4        5        6   
393    2017      85876  9869.0  10156.0  13401.0  11604.0  13748.0  13643.0  \
338    2017       6251     0.0      0.0      0.0      0.0      0.0      0.0   
339    2017       6252     0.0      0.0      0.0      0.0      0.0      0.0   
264    2017       3816     0.0      0.0     12.0     12.0     47.0     44.0   
426    2017      97946  7344.0   6237.0   7825.0   7617.0   7659.0   8163.0   
278    2017       4020     0.0      0.0      1.0     20.0     26.0     49.0   
281    2017       4030     0.0      0.0      4.0     12.0     17.0      8.0   
344    2017       6258     0.0      0.0      0.0      0.0      0.0      0.0   
279    2017       4022     0.0      0.0      1.0     50.0     77.0     82.0   
273    2017       4002     0.0      0.0     21.0     34.0     34.0     40.0   
417    2017      97759  8163.0   7318.0   8219.0   7749.0   7689.0   8017.0   
429    2017      97997  5212.0   4838.0   6007.0   5517.0   5783.0   6161.0   
274    2017       4012     0.0      0.0      0.0     18.0     40.0     37.0   
212    2017       3019     0.0      0.0      0.0      0.0      0.0      0.0   
44     2017     124721  2574.0   2963.0   3502.0   2817.0   2998.0   3348.0   
41     2017     123888  6152.0   6651.0  12303.0  10001.0  10806.0  10125.0   
424    2017      97912  7209.0   6858.0   9239.0   8319.0   8317.0   9129.0   
75     2017       1417  2860.0   2596.0   2951.0   2644.0   2706.0   2489.0   
236    2017       3432  8445.0   7029.0   4434.0   4653.0   4340.0   4641.0   
341    2017       6254     0.0      0.0      0.0      0.0      0.0      0.0   

MONTH        7        8        9       10       11       12  Oct-Nov   
393    18826.0  23338.0  15807.0   5268.0   1200.0    789.0  -4068.0  \
338        0.0      0.0     32.0   5970.0   9847.0  12378.0   3877.0   
339        0.0     10.0     79.0   3075.0   6960.0   7073.0   3885.0   
264       78.0   1173.0   3342.0   6113.0   7895.0   7820.0   1782.0   
426     8821.0   8367.0   9757.0   8774.0  12071.0  10569.0   3297.0   
278      122.0    618.0    794.0   2007.0   4190.0   4150.0   2183.0   
281       62.0    544.0    548.0   2459.0   3836.0   3376.0   1377.0   
344        0.0     16.0    131.0   2082.0   3403.0   3140.0   1321.0   
279      166.0    657.0   1881.0   3047.0   4550.0   4009.0   1503.0   
273      180.0    902.0   1039.0   2155.0   3392.0   3418.0   1237.0   
417     9362.0   8580.0   8995.0   8744.0  10475.0   9114.0   1731.0   
429     6367.0   5750.0   6130.0   7093.0   8179.0   6976.0   1086.0   
274       62.0    303.0    402.0   1308.0   2304.0   2907.0    996.0   
212        0.0    306.0   1394.0   2665.0   3474.0   3915.0    809.0   
44      4552.0   4499.0   4389.0   3368.0   2439.0   1377.0   -929.0   
41     12770.0  13273.0  11226.0  11556.0   9956.0   8095.0  -1600.0   
424     9493.0   8201.0   9110.0   9872.0  10899.0   9238.0   1027.0   
75      2595.0   2111.0   2397.0   3700.0   4314.0   4449.0    614.0   
236        0.0      0.0    738.0   2816.0   2956.0   4047.0    140.0   
341        0.0     83.0    289.0   1374.0   2003.0   2007.0    629.0   

MONTH  Sept-Oct  Perc_Oct-Nov  Perc_Sept-Oct  AvgSeptOct  Nov-AvgSeptOct   
393    -10539.0         -77.2          -66.7     10537.5         -9337.5  \
338      5938.0          64.9        18556.2      3001.0          6846.0   
339      2996.0         126.3         3792.4      1577.0          5383.0   
264      2771.0          29.2           82.9      4727.5          3167.5   
426      -983.0          37.6          -10.1      9265.5          2805.5   
278      1213.0         108.8          152.8      1400.5          2789.5   
281      1911.0          56.0          348.7      1503.5          2332.5   
344      1951.0          63.4         1489.3      1106.5          2296.5   
279      1166.0          49.3           62.0      2464.0          2086.0   
273      1116.0          57.4          107.4      1597.0          1

In [27]:
# Group the rows by ARTICLE_ID and month of DATE, and sum the UNITS values
grouped_sales = sales.groupby(['ARTICLE_ID', pd.Grouper(key='DATE', freq='M')])['UNITS'].sum().reset_index()

# Create new month and year columns
grouped_sales['MONTH'] = grouped_sales['DATE'].dt.month
grouped_sales['YEAR'] = grouped_sales['DATE'].dt.year
grouped_sales = grouped_sales.drop(["DATE"],axis=1)

# Filter for 2016
grouped_sales2016 = grouped_sales[grouped_sales['YEAR'] == 2016]
sales_pivoted2016 = grouped_sales2016.pivot(index=['YEAR', 'ARTICLE_ID'], columns='MONTH', values='UNITS').reset_index()
sales_pivoted2016[5:11]

# Add comparison columns
sales_pivoted2016['Oct-Nov'] = sales_pivoted2016[11]-sales_pivoted2016[10]
sales_pivoted2016['Sept-Oct'] = sales_pivoted2016[10]-sales_pivoted2016[9]
sales_pivoted2016['Perc_Oct-Nov'] = round((sales_pivoted2016[11]-sales_pivoted2016[10])/sales_pivoted2016[10]*100,1)
sales_pivoted2016['Perc_Sept-Oct'] = round((sales_pivoted2016[10]-sales_pivoted2016[9])/sales_pivoted2016[9]*100,1)
sales_pivoted2016['AvgSeptOct'] = sales_pivoted2016[[9, 10]].apply(lambda x: x.mean(), axis=1)
sales_pivoted2016['Nov-AvgSeptOct'] = sales_pivoted2016[11]-sales_pivoted2016['AvgSeptOct']
sales_pivoted2016['Abs(Nov-AvgSeptOct)'] = abs(sales_pivoted2016['Nov-AvgSeptOct'])
sales_pivoted2016[5:11]


MONTH  YEAR ARTICLE_ID       1       2      3      4      5       6      7   
5      2016     108860   398.0   385.0  369.0  287.0  327.0   416.0  284.0  \
6      2016     108894   864.0   729.0  652.0  662.0  830.0  1023.0  866.0   
7      2016     108911   743.0   609.0  650.0  491.0  616.0   689.0  548.0   
8      2016      11187  1086.0  1050.0  751.0  934.0  745.0   879.0  908.0   
9      2016     112345   881.0   619.0  409.0  498.0  589.0   542.0  933.0   
10     2016        113     0.0     0.0    0.0    0.0    0.0     0.0    0.0   

MONTH      8       9      10      11      12  Oct-Nov  Sept-Oct  Perc_Oct-Nov   
5      462.0   541.0   553.0   528.0   545.0    -25.0      12.0          -4.5  \
6      906.0  1281.0  1188.0  1265.0  1548.0     77.0     -93.0           6.5   
7      582.0   813.0   724.0   696.0   860.0    -28.0     -89.0          -3.9   
8      802.0   808.0   733.0   757.0   706.0     24.0     -75.0           3.3   
9      995.0   989.0  1102.0  1376.0  1444.0    274.0     113.0          24.9   
10       0.0     0.0     0.0     0.0     0.0      0.0       0.0           NaN   

MONTH  Perc_Sept-Oct  AvgSeptOct  Nov-AvgSeptOct  Abs(Nov-AvgSeptOct)  
5                2.2       547.0           -19.0                 19.0  
6               -7.3      1234.5            30.5                 30.5  
7              -10.9       768.5           -72.5                 72.5  
8               -9.3       770.5           -13.5                 13.5  
9               11.4      1045.5           330.5                330.5  
10               NaN         0.0             0.0                  0.0

In [28]:
sales_pivoted2016.sort_values(['Abs(Nov-AvgSeptOct)'], ascending=False).head(20)

MONTH  YEAR ARTICLE_ID       1       2       3       4       5        6   
384    2016        832  3468.0  3572.0  4011.0  3571.0  3199.0   3267.0  \
312    2016       4971  1293.0  1443.0  1424.0  1274.0  1165.0    876.0   
187    2016      23240     0.0     0.0   705.0  1918.0  3172.0   3264.0   
393    2016      85876  5173.0  4992.0  5919.0  5200.0  4926.0   6192.0   
348    2016       6505     0.0     0.0     0.0     0.0     0.0      0.0   
382    2016        830  2078.0  2108.0  1898.0  1846.0  1896.0   1822.0   
135    2016      15556     0.0     0.0   181.0   303.0   631.0    910.0   
374    2016      81303  1252.0   981.0  1237.0  1702.0  1452.0   1211.0   
316    2016      53967  3395.0  3682.0  3602.0  4052.0  3758.0   5072.0   
131    2016      15471     0.0     0.0   127.0   289.0   865.0   1034.0   
126    2016      15369     0.0     0.0   124.0   437.0   904.0   1302.0   
107    2016      14944     0.0   860.0  3942.0  4635.0  5481.0   3558.0   
160    2016     189769  2775.0  2725.0  2921.0  3138.0  3667.0   3861.0   
82     2016       1427     0.0     0.0     0.0     0.0     0.0     74.0   
424    2016      97912  8335.0  8175.0  9305.0  8466.0  9196.0  10182.0   
3      2016     106497  2746.0  2770.0  3034.0  3376.0  2645.0   2289.0   
120    2016      15199     0.0     0.0    26.0   618.0  2217.0   2849.0   
124    2016      15335     0.0   441.0  1927.0  2170.0  3034.0   3263.0   
378    2016        826  6368.0  5729.0  5711.0  5813.0  5458.0   5527.0   
411    2016      97623  4198.0  4094.0  5037.0  4663.0  4105.0   4306.0   

MONTH        7        8        9      10       11       12  Oct-Nov  Sept-Oct   
384     3319.0   3595.0   5250.0  7416.0  10280.0   3611.0   2864.0    2166.0  \
312      822.0    931.0   1220.0  1625.0   4972.0   5412.0   3347.0     405.0   
187     3319.0   2342.0   2572.0  2920.0   5466.0   5009.0   2546.0     348.0   
393     9414.0  10163.0  10246.0  9906.0  12119.0  12335.0   2213.0    -340.0   
348        0.0      0.0      0.0   542.0   1570.0   1948.0   1028.0     542.0   
382     1851.0   1951.0   2393.0  2630.0   3805.0   2611.0   1175.0     237.0   
135      761.0    884.0   1319.0  1390.0   2610.0   2723.0   1220.0      71.0   
374     1583.0   2084.0   1766.0  2069.0    776.0   1163.0  -1293.0     303.0   
316     5491.0   5174.0   5237.0  5675.0   4453.0   3993.0  -1222.0     438.0   
131     1001.0   1142.0   1436.0  1243.0   2333.0   1979.0   1090.0    -193.0   
126     1806.0   1611.0   1666.0  1885.0   2710.0   2685.0    825.0     219.0   
107     5489.0   4949.0   5264.0  4087.0   5609.0   5404.0   1522.0   -1177.0   
160     4260.0   5235.0   5700.0  4630.0   4291.0   4072.0   -339.0   -1070.0   
82      1012.0   1315.0   1781.0  1529.0   2500.0   2620.0    971.0    -252.0   
424    10017.0   8272.0   9870.0  9406.0   8795.0   9807.0   -611.0    -464.0   
3       2828.0   3382.0   3299.0  4442.0   3060.0   3217.0  -1382.0    1143.0   
120     3277.0   2577.0   2938.0  3014.0   3740.0   3732.0    726.0      76.0   
124     3283.0   2013.0   2323.0  2362.0   3060.0   3263.0    698.0      39.0   
378     5305.0   5939.0   5712.0  4836.0   4559.0   5030.0   -277.0    -876.0   
411     4210.0   3892.0   4678.0  4457.0   3879.0   4305.0   -578.0    -221.0   

MONTH  Perc_Oct-Nov  Perc_Sept-Oct  AvgSeptOct  Nov-AvgSeptOct   
384            38.6           41.3      6333.0          3947.0  \
312           206.0           33.2      1422.5          3549.5   
187            87.2           13.5      2746.0          2720.0   
393            22.3           -3.3     10076.0          2043.0   
348           189.7            inf       271.0          1299.0   
382            44.7            9.9      2511.5          1293.5   
135            87.8            5.4      1354.5          1255.5   
374           -62.5           17.2      1917.5         -1141.5   
316           -21.5            8.4      5456.0         -1003.0   
131            87.7          -13.4      1339.5           993.5

In [29]:
# Group the rows by ARTICLE_ID and month of DATE, and sum the UNITS values
grouped_sales = sales.groupby(['ARTICLE_ID', pd.Grouper(key='DATE', freq='M')])['UNITS'].sum().reset_index()

# Create new month and year columns
grouped_sales['MONTH'] = grouped_sales['DATE'].dt.month
grouped_sales['YEAR'] = grouped_sales['DATE'].dt.year
grouped_sales = grouped_sales.drop(["DATE"],axis=1)

# Filter for 2015
grouped_sales2015 = grouped_sales[grouped_sales['YEAR'] == 2015]
sales_pivoted2015 = grouped_sales2015.pivot(index=['YEAR', 'ARTICLE_ID'], columns='MONTH', values='UNITS').reset_index()

# Add comparison columns
sales_pivoted2015['Oct-Nov'] = sales_pivoted2015[11]-sales_pivoted2015[10]
sales_pivoted2015['Sept-Oct'] = sales_pivoted2015[10]-sales_pivoted2015[9]
sales_pivoted2015['Perc_Oct-Nov'] = round((sales_pivoted2015[11]-sales_pivoted2015[10])/sales_pivoted2015[10]*100,1)
sales_pivoted2015['Perc_Sept-Oct'] = round((sales_pivoted2015[10]-sales_pivoted2015[9])/sales_pivoted2015[9]*100,1)
sales_pivoted2015['AvgSeptOct'] = sales_pivoted2015[[9, 10]].apply(lambda x: x.mean(), axis=1)
sales_pivoted2015['Nov-AvgSeptOct'] = sales_pivoted2015[11]-sales_pivoted2015['AvgSeptOct']
sales_pivoted2015['Abs(Nov-AvgSeptOct)'] = abs(sales_pivoted2015['Nov-AvgSeptOct'])


In [30]:
sales_pivoted2015.sort_values(['Abs(Nov-AvgSeptOct)'], ascending=False).head(20)

MONTH  YEAR ARTICLE_ID        4        5        6        7        8        9   
314    2015       5148      0.0      0.0      0.0      0.0      0.0    862.0  \
417    2015      97759  10183.0  11861.0  11774.0   8448.0   8527.0   8884.0   
313    2015       5130      0.0      0.0      0.0      0.0      0.0    692.0   
426    2015      97946   5657.0   7102.0   7610.0   7553.0   6195.0   6290.0   
424    2015      97912   8125.0  11819.0  11616.0  10019.0   9574.0   9105.0   
50     2015     136094   4239.0   5558.0   5941.0   5020.0   4457.0   4078.0   
37     2015     122817   2278.0   2718.0   2720.0   2653.0   3415.0   4231.0   
315    2015       5151      0.0      0.0      0.0      0.0      0.0    321.0   
239    2015       3438   3244.0   2680.0   3064.0   4442.0   5140.0   3766.0   
3      2015     106497   4011.0   3735.0   4022.0   3808.0   4941.0   3136.0   
416    2015      97742   4446.0   5226.0   5546.0   5458.0   4728.0   3802.0   
393    2015      85876   5516.0   6820.0   4845.0   3797.0   3847.0   4959.0   
311    2015       4969      0.0      0.0      0.0      0.0      0.0    317.0   
144    2015      17902   1203.0   1432.0   1659.0   1620.0   1614.0   1576.0   
420    2015      97827   3574.0   3927.0   4269.0   4466.0   3654.0   3071.0   
427    2015      97963   3488.0   4392.0   4558.0   4202.0   3466.0   3136.0   
41     2015     123888  10163.0  11102.0  10850.0  11018.0  11118.0   9692.0   
385    2015        833   3952.0   4710.0   4529.0   3956.0   3583.0   4308.0   
423    2015      97895   8700.0   9511.0   9870.0   9794.0   8560.0   7431.0   
16     2015     114674   8490.0  13296.0  12618.0  12163.0  12035.0  10204.0   

MONTH       10       11       12  Oct-Nov  Sept-Oct  Perc_Oct-Nov   
314     3849.0   5151.0   4276.0   1302.0    2987.0          33.8  \
417    10551.0  11865.0  10094.0   1314.0    1667.0          12.5   
313     3216.0   3997.0   3514.0    781.0    2524.0          24.3   
426     7183.0   8614.0   7977.0   1431.0     893.0          19.9   
424     9734.0  11123.0   9303.0   1389.0     629.0          14.3   
50      4947.0   5840.0   5094.0    893.0     869.0          18.1   
37      3296.0   2507.0   1383.0   -789.0    -935.0         -23.9   
315     2047.0   2401.0   1933.0    354.0    1726.0          17.3   
239     2460.0   4269.0   4394.0   1809.0   -1306.0          73.5   
3       2640.0   3981.0   3034.0   1341.0    -496.0          50.8   
416     4323.0   5120.0   4505.0    797.0     521.0          18.4   
393     6049.0   6522.0   6309.0    473.0    1090.0           7.8   
311     1822.0   2034.0   1879.0    212.0    1505.0          11.6   
144     1741.0   2607.0   1897.0    866.0     165.0          49.7   
420     3440.0   4196.0   3975.0    756.0     369.0          22.0   
427     3277.0   4147.0   3668.0    870.0     141.0          26.5   
41      9670.0   8766.0   8187.0   -904.0     -22.0          -9.3   
385     4927.0   3708.0   3314.0  -1219.0     619.0         -24.7   
423     7538.0   8380.0   7973.0    842.0     107.0          11.2   
16     10260.0   9382.0   8479.0   -878.0      56.0          -8.6   

MONTH  Perc_Sept-Oct  AvgSeptOct  Nov-AvgSeptOct  Abs(Nov-AvgSeptOct)  
314            346.5      2355.5          2795.5               2795.5  
417             18.8      9717.5          2147.5               2147.5  
313            364.7      1954.0          2043.0               2043.0  
426             14.2      6736.5          1877.5               1877.5  
424              6.9      9419.5          1703.5               1703.5  
50              21.3      4512.5          1327.5               1327.5  
37             -22.1      3763.5         -1256.5               1256.5  
315            537.7      1184.0          1217.0               1217.0  
239            -34.7      3113.0          1156.0               1156.0  
3              -15.8      2888.0          1093.0               1093.0  
416             13.7      4062.5          1057.5               1057.5  
393  

In [37]:
# Get top 20 most impactful articles (from above) for each year
x = sales_pivoted2017.sort_values(['Abs(Nov-AvgSeptOct)'], ascending=False)['ARTICLE_ID'].head(20)
y = sales_pivoted2016.sort_values(['Abs(Nov-AvgSeptOct)'], ascending=False)['ARTICLE_ID'].head(20)
z = sales_pivoted2015.sort_values(['Abs(Nov-AvgSeptOct)'], ascending=False)['ARTICLE_ID'].head(20)

x, y, z = set(x), set(y), set(z)

print("Articles with worst Nov predictions from 2015-2017: ", x.intersection(y,z))
print("Articles with worst Nov predictions from 2016-2017: ", x.intersection(y))

Articles with worst Nov predictions from 2015-2017:  {'97912', '85876'}
Articles with worst Nov predictions from 2016-2017:  {'97912', '85876'}


In [45]:
sales_pivoted[(sales_pivoted["ARTICLE_ID"] == "97912")] # | (sales_pivoted["ARTICLE_ID"] == "85876")]

MONTH  YEAR ARTICLE_ID       1       2       3       4        5        6   
424    2015      97912     NaN     NaN     NaN  8125.0  11819.0  11616.0  \
859    2016      97912  8335.0  8175.0  9305.0  8466.0   9196.0  10182.0   
1294   2017      97912  7209.0  6858.0  9239.0  8319.0   8317.0   9129.0   
1729   2018      97912  7579.0  7251.0  8030.0  6842.0   7689.0   8930.0   

MONTH        7       8       9      10       11      12  
424    10019.0  9574.0  9105.0  9734.0  11123.0  9303.0  
859    10017.0  8272.0  9870.0  9406.0   8795.0  9807.0  
1294    9493.0  8201.0  9110.0  9872.0  10899.0  9238.0  
1729    8622.0  8113.0  8384.0  8175.0      NaN     NaN

In [51]:
# get units sold from Jan-Oct each year for article 97912
grouped_sales[(grouped_sales["ARTICLE_ID"] == "97912") & (grouped_sales["MONTH"] != 11) & (grouped_sales["MONTH"] != 12)][["YEAR","UNITS"]].groupby(["YEAR"]).sum()


UNITS
YEAR         
2015  69992.0
2016  91224.0
2017  85747.0
2018  79615.0

The above suggests that Sept-Oct cannot reliably predict two articles, 97912 and 85876 for November sales. 

In fact, 85876 shows dwindling sales throughout 2018. It would be reasonable to predict fewer than 12 units for this article given the above. 

Article 97912 consistently has a high sales volume. In 2015 and 2017, this article sold more units in Nov than in Sept or Oct, but the opposite was true in 2016. Meanwhile, the article has fewer sales in 2018 when compared overall to 2016 & 2017.

Questions: 

- Why such a big jump in sales for article 106497 in November 2015, then October 2016 and pretty big in September 2017?

- Similarly, for articles 106650 , September is a much better predictor for November than October is from 2015-2017.

- 98150 seems to make big leaps at the end of the year

- What happens if focus on articles with highest volumes of transactions? 

- 3438 - how did this article have almost no sales in some months in 2017? 

## Article Drill Down

In [66]:
# 3438 units sold in August 2017
get_total_units_sold(sales, year=2017, month=8, article_id=3438)

4.0